### DISCUSSION TITLE
Animate Node transform/How to create Node Animations

### DISCUSSION THREAD

**Discussion by paricleu** [sceneview/sceneview#482] [Q&A]
I've ran into a problem animating Node transforms.
Transforms without animations work well e.g. `node.transform(position, rotation, scale)` and even smooth transformation are fine e.g. 
```
node.transform(position, rotation, scale, true, 1.0)
```
Even tho it is nowhere documented what the "smoothSpeed" is referencing to. Is is seconds? It does not seem to be.

What I am trying to achieve is animating a node and also be able to stop the animation if it is not finished yet.
I came across 
```
    fun animatePositions(vararg positions: Position): ObjectAnimator =
        NodeAnimator.ofPosition(this, *positions)

    fun animateQuaternions(vararg quaternions: Quaternion): ObjectAnimator =
        NodeAnimator.ofQuaternion(this, *quaternions)

    fun animateRotations(vararg rotations: Rotation): ObjectAnimator =
        NodeAnimator.ofRotation(this, *rotations)

    fun animateScales(vararg scales: Scale): ObjectAnimator =
        NodeAnimator.ofScale(this, *scales)

    fun animateTransforms(vararg transforms: Transform): AnimatorSet =
        NodeAnimator.ofTransform(this, *transforms)
```
in Node.kt but I have no clue how to use it.
`node.animateRotations(rotation).start()` does nothing. 

How can I create a node animation so it is cancelable?

**Comment by punkesta** [comment]
Have solution?

**Comment by ofreye** [comment]
Would be interested too...

**Comment by ThomasGorisse** [comment]
Hi,
The `NodeAnimator` is missing an update after the node transform applying on each frame remove.
We will fix it in next release.
In the meantime, you can use your own implementation with a `ValueAnimator`:
```kotlin
ValueAnimator.ofFloat(0f, 360f).apply {
    setDuration(10000.0f)
    repeatCount = INFINITE
    addUpdateListener {
        val value = it.animatedValue as Float
        node.rotation = Rotation(y = -value)
    }
}.start()
```

**Comment by ThomasGorisse** [comment]
Linked ticket: #534

**Comment by idonahum1** [comment]
Any update regarding this?

**Comment by ThomasGorisse** [comment]
In v4.0.0-rc.1, you can animate node transforms using Compose state + animation APIs:

```kotlin
var targetPosition by remember { mutableStateOf(Position(0f, 0f, 0f)) }

// Use Compose animation
val animatedPosition by animateValueAsState(targetPosition)

SceneView {
    ModelNode(
        modelInstance = model,
        position = animatedPosition
    )
}
```

For GLB-embedded animations, use the Filament `Animator`:
```kotlin
model?.animator?.playAnimation(0)
```

The `sceneview-core` KMP module also provides Spring, Property, and Interpolation animation utilities for more advanced programmatic animations.
